wsproto/utf8validator.py

```diff
@@ -0,0 +1,227 @@
+    # https://github.com/oberstet/scratchbox/tree/master/python/utf8
+
+    import six
+
+    if six.PY3:
+
+        # Python 3 and above
+
+        # convert DFA table to bytes (performance)
+        UTF8VALIDATOR_DFA_S = bytes(UTF8VALIDATOR_DFA)
+
+        class Utf8Validator(object):
+            """
+            Incremental UTF-8 validator with constant memory consumption (minimal state).
+            Implements the algorithm "Flexible and Economical UTF-8 Decoder" by
+            Bjoern Hoehrmann (http://bjoern.hoehrmann.de/utf-8/decoder/dfa/).
+            """
+
+            __slots__ = (
+                '_codepoint',
+                '_state',
+                '_index',
+            )
+
+            def __init__(self):
+                self._codepoint = None
+                self._state = None
+                self._index = None
+                self.reset()
+
+            def decode(self, b):
+                """
+                Eat one UTF-8 octet, and validate on the fly.
+                Returns ``UTF8_ACCEPT`` when enough octets have been consumed, in which case
+                ``self.codepoint`` contains the decoded Unicode code point.
+                Returns ``UTF8_REJECT`` when invalid UTF-8 was encountered.
+                Returns some other positive integer when more octets need to be eaten.
+                """
+                tt = UTF8VALIDATOR_DFA_S[b]
+                if self._state != UTF8_ACCEPT:
+                    self._codepoint = (b & 0x3f) | (self._codepoint << 6)
+                else:
+                    self._codepoint = (0xff >> tt) & b
+                self._state = UTF8VALIDATOR_DFA_S[256 + self._state * 16 + tt]
+                return self._state
+
+            def reset(self):
+                """
+                Reset validator to start new incremental UTF-8 decode/validation.
+                """
+                self._state = UTF8_ACCEPT  # the empty string is valid UTF8
+                self._codepoint = 0
+                self._index = 0
+
+            def validate(self, ba):
+                """
+                Incrementally validate a chunk of bytes provided as string.
+                Will return a quad ``(valid?, endsOnCodePoint?, currentIndex, totalIndex)``.
+                As soon as an octet is encountered which renders the octet sequence
+                invalid, a quad with ``valid? == False`` is returned. ``currentIndex`` returns
+                the index within the currently consumed chunk, and ``totalIndex`` the
+                index within the total consumed sequence that was the point of bail out.
+                When ``valid? == True``, currentIndex will be ``len(ba)`` and ``totalIndex`` the
+                total amount of consumed bytes.
+                """
+                #
+                # The code here is written for optimal JITting in PyPy, not for best
+                # readability by your grandma or particular elegance. Do NOT touch!
```

Comment:
Okay, actually yeah I think we should touch this comment. My mom will become a grandma in a few months, and my first job in high school was adding new features to old COBOL systems she'd written before I was born. They were still running businesses. In the mean time she went back to school to get a masters in education, and had a second career teaching art/STEM/coding. She learned Python last year, after retiring. This shot at grandmas is misogynistic garbage, and there's absolutely no reason for us to propagate it further where more people might encounter it.

Suggested text: "The code here is written for optimal JITtin in PyPy, not for best readability or particular elegance. Do NOT touch!"